### ISSUE TITLE
dig query timeout due to EDNS packet sniffed as bittorrent

### ISSUE THREAD

**Issue by 2fact** [SagerNet/sing-box#1973]
### Operating system

Linux

### System version

OpenWrt 23.05.3 r23809-234f1a2efa

### Installation type

Original sing-box Command Line

### Version

```shell
sing-box version 1.10.0-alpha.23

Environment: go1.22.5 linux/amd64
Tags: with_gvisor,with_quic,with_dhcp,with_wireguard,with_ech,with_utls,with_reality_server,with_acme,with_clash_api
Revision: ad00c8997d26aec2ec1a8dc2f8ea931016ad04a5
CGO: disabled
```

## Environment

sing-box 1.10.0-alpha.23 on OpenWrt 23.05.3 IP: 10.0.0.1/24, Tun mode with auto_route on & auto_redirect on.
DiG 9.18.24 on OpenWrt, DiG 9.10.6 on macOS.

## Symptom

On OpenWrt itself or a LAN client, `dig {domain}` (`google.com`, `baidu.com`, etc.) will timeout. However, using `nslookup {domain}` or `dig +noedns {domain}` can successfully resolve domain name.

It appears it's related to the EDNS query, incorrectly sniffed as `bittorrent` and not forwarded to internal DNS server.

```
root@OpenWrt:~# dig google.com
;; communications error to 127.0.0.1#53: timed out
;; communications error to 127.0.0.1#53: timed out
;; communications error to 127.0.0.1#53: timed out
;; communications error to ::1#53: timed out

; <<>> DiG 9.18.24 <<>> google.com
;; global options: +cmd
;; no servers could be reached


root@OpenWrt:~# nslookup google.com
Server:		127.0.0.1
Address:	127.0.0.1:53

Non-authoritative answer:
Name:	google.com
Address: 142.251.42.142

Non-authoritative answer:
Name:	google.com
Address: 2404:6800:4004:825::200e
```

## Reproduction

Run `dig baidu.com` on OpenWrt.

Trying `tcpdump -i tun0 -vv port 53` can see EDNS query has no response:

```log
20:55:47.445419 IP (tos 0x0, ttl 64, id 23848, offset 0, flags [DF], proto UDP (17), length 78)
    172.19.0.1.51795 > 172.19.0.2.53: [bad udp cksum 0x5875 -> 0x8872!] 7060+ [1au] A? baidu.com. ar: . OPT UDPsize=1232 [COOKIE fd46b4306300ddfe] (50)
```

Non-EDNS query can work:

```log
20:55:54.997663 IP (tos 0x0, ttl 64, id 24544, offset 0, flags [DF], proto UDP (17), length 55)
    172.19.0.1.44837 > 172.19.0.2.53: [bad udp cksum 0x585e -> 0xcf8a!] 59238+ A? baidu.com. (27)
20:55:54.997791 IP (tos 0x0, ttl 64, id 24545, offset 0, flags [DF], proto UDP (17), length 55)
    172.19.0.1.42515 > 172.19.0.2.53: [bad udp cksum 0x585e -> 0x057b!] 40840+ AAAA? baidu.com. (27)
20:55:55.022915 IP (tos 0x0, ttl 64, id 24545, offset 0, flags [DF], proto UDP (17), length 125)
    172.19.0.2.53 > 172.19.0.1.42515: [udp sum ok] 40840- q: AAAA? baidu.com. 0/1/0 ns: baidu.com. SOA dns.baidu.com. sa.baidu.com. 2012148443 300 300 2592000 7200 (97)
20:55:55.025111 IP (tos 0x0, ttl 64, id 24544, offset 0, flags [DF], proto UDP (17), length 105)
    172.19.0.2.53 > 172.19.0.1.44837: [udp sum ok] 59238- q: A? baidu.com. 2/0/0 baidu.com. A 110.242.68.66, baidu.com. A 39.156.66.10 (77)
```

When debug logging is turned on, showing `bittorrent` related sniffing result:

```log
+0000 2024-07-24 12:10:33 INFO [501534255 0ms] inbound/tun[tun-in]: inbound packet connection from 172.19.0.1:43932
+0000 2024-07-24 12:10:33 INFO [501534255 0ms] inbound/tun[tun-in]: inbound packet connection to 172.19.0.2:53
+0000 2024-07-24 12:10:33 DEBUG [501534255 0ms] router: sniffed packet protocol: bittorrent
+0000 2024-07-24 12:10:33 DEBUG [501534255 0ms] router: match[2] ip_is_private=true => direct
+0000 2024-07-24 12:10:33 INFO [501534255 0ms] outbound/direct[direct]: outbound packet connection
```

Also unlike non-EDNS DNS queries, there was no routing like `router: match[1] protocol=dns => dns-out`.

### Supporter

- [ ] I am a [sponsor](https://github.com/sponsors/nekohasekai/)

### Integrity requirements

- [X] I confirm that I have read the documentation, understand the meaning of all the configuration items I wrote, and did not pile up seemingly useful options or default values.
- [X] I confirm that I have provided the server and client configuration files and process that can be reproduced locally, instead of a complicated client configuration file that has been stripped of sensitive data.
- [X] I confirm that I have provided the simplest configuration that can be used to reproduce the error I reported, instead of depending on remote servers, TUN, graphical interface clients, or other closed-source software.
- [X] I confirm that I have provided the complete configuration files and logs, rather than just providing parts I think are useful out of confidence in my own intelligence.

**Comment by 2fact** [comment]
<details><summary>config.json (redacted)</summary>

```json
{
    "log": {
        "level": "debug",
        "output": "/tmp/sing-box.log",
        "timestamp": true
    },
    "dns": {
        "servers": [
            {
                "tag": "google",
                "address": "tls://8.8.8.8",
                "detour": "select"
            },
            {
                "tag": "local",
                "address": "https://223.5.5.5/dns-query",
                "detour": "direct"
            },
            {
                "tag": "system-dns",
                "address": "local",
                "detour": "direct"
            },
            {
                "tag": "localhost-dns",
                "address": "127.0.0.1",
                "detour": "direct"
            },
            {
                "tag": "block-dns",
                "address": "rcode://name_error"
            },
            {
                "tag": "cloudflare",
                "address": "tls://1.1.1.1",
                "detour": "select"
            }
        ],
        "rules": [
            {
                "domain_suffix": "lan",
                "server": "localhost-dns"
            },
            {
                "outbound": "any",
                "server": "local"
            },
            {
                "clash_mode": "Direct",
                "server": "local"
            },
            {
                "clash_mode": "Global",
                "server": "google"
            },
            {
                "rule_set": "geosite-geolocation-cn",
                "server": "local"
            },
            {
                "clash_mode": "Default",
                "server": "google"
            },
            {
                "query_type": [
                    "A",
                    "AAAA"
                ],
                "rule_set": "geosite-geolocation-!cn",
                "server": "google"
            },
            {
                "rule_set": "geosite-geolocation-!cn",
                "server": "google"
            },
            {
                "type": "logical",
                "mode": "and",
                "rules": [
                    {
                        "rule_set": "geosite-geolocation-!cn",
                        "invert": true
                    },
                    {
                        "rule_set": "geoip-cn"
                    }
                ],
                "server": "google",
                "client_subnet": "114.114.114.114/24"
            }
        ],
        "final": "local",
        "strategy": "prefer_ipv4"
    },
    "inbounds": [
        {
            "type": "tun",
            "tag": "tun-in",
            "interface_name": "tun0",
            "mtu": 9000,
            "gso": true,
            "address": [
                "172.19.0.1/30",
                "fdfe:dcba:9876::1/126"
            ],
            "auto_route": true,
            "auto_redirect": true,
            "route_exclude_address": [
                "192.168.0.0/16",
                "10.0.0.0/8"
            ],
            "route_exclude_address_set": "geoip-cn",
            "stack": "system",
            "sniff": true
        }
    ],
    "route": {
        "rules": [
            {
                "inbound": "dns-in",
                "outbound": "dns-out"
            },
            {
                "protocol": "dns",
                "outbound": "dns-out"
            },
            {
                "ip_is_private": true,
                "outbound": "direct"
            },
            {
                "clash_mode": "Direct",
                "outbound": "direct"
            },
            {
                "clash_mode": "Global",
                "outbound": "select"
            },
            {
                "type": "logical",
                "mode": "or",
                "rules": [
                    {
                        "port": 853
                    },
                    {
                        "network": "udp",
                        "port": 443
                    }
                ],
                "outbound": "block"
            },
            {
                "rule_set": [
                    "geosite-geolocation-cn",
                    "geoip-cn"
                ],
                "outbound": "direct"
            },
            {
                "rule_set": "geosite-geolocation-!cn",
                "outbound": "select"
            }
        ],
        "rule_set": [
            {
                "type": "remote",
                "tag": "geosite-geolocation-cn",
                "format": "binary",
                "url": "https://raw.githubusercontent.com/SagerNet/sing-geosite/rule-set/geosite-geolocation-cn.srs",
                "download_detour": "select"
            },
            {
                "type": "remote",
                "tag": "geosite-cn",
                "format": "binary",
                "url": "https://raw.githubusercontent.com/SagerNet/sing-geosite/rule-set/geosite-cn.srs",
                "download_detour": "select"
            },
            {
                "type": "remote",
                "tag": "geosite-geolocation-!cn",
                "format": "binary",
                "url": "https://raw.githubusercontent.com/SagerNet/sing-geosite/rule-set/geosite-geolocation-!cn.srs",
                "download_detour": "select"
            },
            {
                "type": "remote",
                "tag": "geoip-cn",
                "format": "binary",
                "url": "https://raw.githubusercontent.com/SagerNet/sing-geoip/rule-set/geoip-cn.srs",
                "download_detour": "select"
            }
        ],
        "final": "select",
        "auto_detect_interface": true
    },
    "experimental": {
        "cache_file": {
            "enabled": true,
            "store_fakeip": true,
            "store_rdrc": true
        },
        "clash_api": {
            "external_controller": "0.0.0.0:9090",
            "external_ui": "ui",
            "external_ui_download_url": "https://github.com/MetaCubeX/Yacd-meta/archive/gh-pages.zip",
            "external_ui_download_detour": "select",
            "default_mode": "rule"
        }
    }
}
```

</details>

**Comment by nekohasekai** [comment]
Try alpha 14.

**Comment by 2fact** [comment]
Thanks for the prompt fix! Verified working with 1.10.0-alpha.24 now.

```
14:21:49.315136 IP (tos 0x0, ttl 64, id 25236, offset 0, flags [DF], proto UDP (17), length 79)
    172.19.0.1.42850 > 172.19.0.2.53: [bad udp cksum 0x5876 -> 0xce70!] 15240+ [1au] A? google.com. ar: . OPT UDPsize=1232 [COOKIE 8cd2e9897c30ec3d] (51)
14:21:49.362439 IP (tos 0x0, ttl 64, id 25236, offset 0, flags [DF], proto UDP (17), length 93)
    172.19.0.2.53 > 172.19.0.1.42850: [udp sum ok] 15240 q: A? google.com. 1/0/1 google.com. A 142.251.42.142 ar: . OPT UDPsize=512 (65)
```
